.github/workflows/optimize-ci-build-reusable.yml

```diff
@@ -0,0 +1,102 @@
+# This GitHub Actions workflow automates the CI build process for the 'optimize' service.
+# It triggers on a `workflow_call` event and accepts inputs for branch name[required], Java version[optional]
+#
+# It consists of a several steps:
+# 1. Setup: Checks out the branch, defines environment variables, fetches the main branch, reads Java/Maven project info, logs into the Harbor registry, and sets up Maven.
+# 2. Build: Generates a production .tar.gz file using Maven and builds a Docker image with appropriate tags.
+# 3. Upload: Pushes the Docker image to the Harbor registry tagged for the pull request.
+#
+# Environment variables are used to define the image tags used for Docker builds.
+# This workflow is designed to provide a comprehensive, automated CI process that ensures code quality, handles secrets securely,
+# and enables detailed reporting of test results.
+
+name: Optimize CI Reusable
+
+on:
+  workflow_call:
+    inputs:
+      branch:
+        description: "The branch name to be used for the workflow"
+        required: true
+        type: string
+      pushDocker:
+        description: "Whether the built docker images are pushed to camunda registry"
+        type: boolean
+        default: true
+      previewEnv:
+        description: "Defines whether this is a preview environment build or not. If true, the image will be pushed to the Harbor Docker registry."
+        type: boolean
+        default: false
+
+defaults:
+  run:
+    # use bash shell by default to ensure pipefail behavior is the default
+    # see https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#exit-codes-and-error-action-preference
+    shell: bash
+
+jobs:
+  build:
+    name: Build
+    runs-on: ubuntu-latest
+    timeout-minutes: 30
+    steps:
+      # Setup: checkout branch
+      - name: Checkout '${{ inputs.branch }}' branch
+        uses: actions/checkout@v4
+        with:
+          ref: refs/heads/${{ inputs.branch }}
+          fetch-depth: 0 # fetches all history for all branches and tags
+
+      #########################################################################
+      # Setup: define env variables
+      - name: Set GitHub environment variables
+        run: |
+          GIT_COMMIT_HASH=$(git rev-parse ${{ inputs.branch }})
+          echo "GIT_COMMIT_HASH=$GIT_COMMIT_HASH" >> $GITHUB_ENV
+          echo "IMAGE_TAG=registry.camunda.cloud/team-camunda/optimize:pr-$GIT_COMMIT_HASH" >> $GITHUB_ENV
+          if [ "${{ inputs.previewEnv }}" = "true" ]; then
+            echo "IMAGE_TAG=registry.camunda.cloud/team-camunda/optimize:pr-$GIT_COMMIT_HASH" >> $GITHUB_ENV
+          else
+            echo "IMAGE_TAG=registry.camunda.cloud/team-optimize/optimize:pr-$GIT_COMMIT_HASH" >> $GITHUB_ENV
```

Comment:
@andromaqui this logic here was carried over from the Optimize repository and Docker images called `registry.camunda.cloud/team-optimize/optimize:pr*` are only used for the preview environments. There is no need to preserve this logic in the monorepo. Outside of `inputs.previewEnv == true` there is no need to push any Docker image here.

Reply:
I am not very about this, the [code](https://github.com/camunda/camunda/pull/20885/files#diff-9232e03df9f79b5a046a26e4ca02fa62410d4a501a9b79be5b1f577358ef3f4dL64-L99) which previously was on optimize-ci.yml was tagging and pushing this image to `registry.camunda.cloud/team-optimize/optimize:pr`. Do you think this was a mistake?

Reply:
I think it was unused since Docker images tagged with `pr-*` are used only for preview environments, see https://github.com/search?q=repo%3Acamunda%2Fcamunda-optimize+pr-&type=code

Since the preview environments that you are building do not use `registry.camunda.cloud/team-optimize/optimize` but instead use `registry.camunda.cloud/team-camunda/optimize` the old logic is useless, yes.

Reply:
The [optimize-ci workflow](https://github.com/camunda/camunda/blob/main/.github/workflows/optimize-ci.yml#L58) which I attached though is part of the camunda/camunda repo and runs on every pr associated with Optimize changes, is it really unused? In there there image is being tagged with [pr-${{ github.event.pull_request.head.sha }}](https://github.com/camunda/camunda/blob/main/.github/workflows/optimize-ci.yml#L98) I believe.

The [optimize-ci-build-reusable.yml](https://github.com/camunda/camunda/pull/20885/files/56ff44a9f750fdd506fa76d16a8949d936f7c57d#diff-a6380d6c48fac8323a2728ce2b1e095371ee07ab8885fd8af769b4e38f8ddedd) is being used from both the `optimize-ci `and the preview environment workflow

Reply:
The workflow is used for sure but the specific line I commented on contains logic that is unused: namely this to tag Docker images with `registry.camunda.cloud/team-optimize/optimize:pr-*` and push them. Nobody consumes them so pushing them is useless (in fact it costs Camunda money for nothing, not your fault) and if we don't need to push we also don't need that `else` condition to assign that tag.

Reply:
Alrght, thank you for the explanation! That would change the logic which exists currently in the optimize-ci workflow, is this something we need to warn the Optimize team about?

Reply:
As we need to tell them anyways about this bigger change to their pipeline, we should mention this as well 👍🏼